File: projects/gnrcore/packages/test15/webpages/chart/canvas.py
# -*- coding: UTF-8 -*-
# 
"""Canvas"""

class GnrCustomWebPage(object):
    py_requires = "gnrcomponents/testhandler:TestHandlerBase"
    js_requires ='canvas'
    dojo_theme = 'tundra'
    
    def test_1_basic(self, pane):
        """Basic canvas"""
        pane=pane.div(background_color='white',height='400px')
        fb=pane.formBuilder(cols=4,lblpos='T',border_spacing='6px')
        kwargs=dict(width='200px',height='150px',
                                            border='1px solid silver',shadow='2px 2px 4px silver')
        
        fb.canvas(onCreated="""var ctx =widget.getContext('2d');
                                            ctx.fillStyle = "rgb(200,0,0)";
                                            ctx.fillRect (10, 10, 55, 50);

                                            ctx.fillStyle = "rgba(0, 0, 200, 0.5)";
                                            ctx.fillRect (30, 30, 55, 50);""",lbl='Test 1',**kwargs)
                                            
        fb.canvas(onCreated="""var ctx =widget.getContext('2d');
                                           ctx.fillRect(25,25,100,100);
                                           ctx.clearRect(45,45,60,60);
                                           ctx.strokeRect(50,50,50,50);""",lbl='Test 2',**kwargs)
                                           
        fb.canvas(onCreated="""var ctx =widget.getContext('2d');
                                            ctx.beginPath();
                                            ctx.moveTo(75,50);
                                            ctx.lineTo(100,75);
                                            ctx.lineTo(100,25);
                                            ctx.fill();""",lbl='Test 3',**kwargs)
                                            
        fb.canvas(onCreated="""var ctx =widget.getContext('2d');
                                            ctx.beginPath();
                                            ctx.arc(75,75,50,0,Math.PI*2,true); // Outer circle
                                            ctx.moveTo(110,75);
                                            ctx.arc(75,75,35,0,Math.PI,false);   // Mouth (clockwise)
                                            ctx.moveTo(65,65);
                                            ctx.arc(60,65,5,0,Math.PI*2,true);  // Left eye
                                            ctx.moveTo(95,65);
                                            ctx.arc(90,65,5,0,Math.PI*2,true);  // Right eye
                                            ctx.stroke();""",lbl='Test 4',**kwargs)
        fb.canvas(onCreated="""var ctx =widget.getContext('2d');
                                                 ctx.beginPath();
                                                 ctx.moveTo(25,25);
                                                 ctx.lineTo(105,25);
                                                 ctx.lineTo(25,105);
                                                 ctx.fill();
                                                 ctx.beginPath();
                                                 ctx.moveTo(125,125);
                                                 ctx.lineTo(125,45);
                                                 ctx.lineTo(45,125);
                                                 ctx.closePath();
                                                 ctx.stroke();
                                            """,lbl='Test 5',**kwargs)
                                         
        fb.canvas(onCreated="""var ctx =widget.getContext('2d');
                                 r(var i=0;i<4;i++){
                                  for(var j=0;j<3;j++){
                                  ctx.beginPath();
                                  var x              = 25+j*50;               // x coordinate
                                  var y              = 25+i*50;               // y coordinate
                                  var radius         = 20;                    // Arc radius
                                  var startAngle     = 0;                     // Starting point on circle
                                  var endAngle       = Math.PI+(Math.PI*j)/2; // End point on circle
                                  var anticlockwise  = i%2==0 ? false : true; // clockwise or anticlockwise
                                 
                                  ctx.arc(x,y,radius,startAngle,endAngle, anticlockwise);
                                 
                                  if (i>1){
                                      ctx.fill();
                                      } else {
                                      ctx.stroke();
                                      }
                                      }
                                  }""",lbl='Test 6 ',**kwargs)
     

        fb.canvas(onCreated="""var ctx =widget.getContext('2d');
                                    ctx.beginPath();
                                    ctx.moveTo(75,40);
                                    ctx.bezierCurveTo(75,37,70,25,50,25);
                                    ctx.bezierCurveTo(20,25,20,62.5,20,62.5);
                                    ctx.bezierCurveTo(20,80,40,102,75,120);
                                    ctx.bezierCurveTo(110,102,130,80,130,62.5);
                                    ctx.bezierCurveTo(130,62.5,130,25,100,25);
                                    ctx.bezierCurveTo(85,25,75,37,75,40);
                                    ctx.fill()""",lbl='Test 7 ',**kwargs)

    def test_2_special(self, pane):
        """Basic canvas"""
        pane=pane.div(background_color='white',height='600px')
        fb=pane.formBuilder(cols=4,lblpos='T',border_spacing='6px')
        kwargs=dict(width='700px',height='500px',
                                            border='1px solid silver',shadow='2px 2px 4px silver')
        fb.canvas(onCreated="""var ctx =widget.getContext('2d');
                                            var x = 188;
                                            var y = 130;
                                            var width = 200;
                                            var height = 137;
                                            var imageObj = new Image();
                                            imageObj.onload = function() {
                                                ctx.drawImage(imageObj, x, y, width, height);
                                            };
                                            imageObj.src = "http://www.html5canvastutorials.com/demos/assets/darth-vader.jpg";
""",lbl='Test X',**kwargs)




    def test_3_special(self, pane):
        """Basic canvas"""
        pane=pane.div(background_color='white',height='600px')
        fb=pane.formBuilder(cols=4,lblpos='T',border_spacing='6px')
        kwargs=dict(width='700px',height='500px',
                                            border='1px solid silver',shadow='2px 2px 4px silver')
        fb.canvas(onCreated="""var canvas=widget;var context =widget.getContext('2d');
                               var arc=function(ctx,cx,cy,radius,start,end,linew){
                                         ctx.arc(cx, cy, radius,  start * Math.PI,  end * Math.PI, true);
                                         ctx.lineWidth = linew;
                                         ctx.strokeStyle = "black";
                                         ctx.stroke();
                                         }
                                    line=function(ctx,x0,y0,x1,y1,linew){
                                         ctx.moveTo(x0, y0);
                                         ctx.lineTo(x1, y1);
                                         ctx.lineWidth = linew;
                                         ctx.strokeStyle = "black";
                                         ctx.stroke();
                                    }
                                        var w=canvas.width
                                        var h=canvas.height
                                        
                                        arc(context,w/2,h/2,40,0,-0.5,10)

""",lbl='Test X',**kwargs)


    def test_4_special(self, pane):
        """Basic canvas"""
        pane=pane.div(background_color='white',height='700px')
        fb=pane.formBuilder(cols=3,lblpos='T',border_spacing='6px')
        kwargs=dict(width='250px',height='200px',zoom='1',
                                            border='1px solid silver',shadow='2px 2px 4px silver')
        sizes=[50,60,70,80,90,100,110,120,130]
        for size in sizes:
            fb.canvas(onCreated="""gnrcanvas.genropig(widget,{strokeStyle:'rgb(40,40,40)',
                                                              fillStyle:'rgb(0,150,0)',
                                                              size:%i
                                                             })    
                                                          """%size,lbl='Genropig %s'%size,**kwargs)
    def test_5_special(self, pane):
         pane=pane.div(background_color='white',width='800px')
         pane.canvas(height='500px',width='700px',onCreated="""gnrcanvas.genropig(widget,{strokeStyle:'rgb(40,40,40)',
                                                              fillStyle:'rgb(0,150,0)',
                                                              size:400
                                                             })    
                                                          """)
        
         

File: projects/gnrcore/packages/test15/resources/canvas.js
var gnrcanvas={

    muzzleline:function(ctx,cx,cy,w,h,r,b,k,rm){
        var r=r +b-rm ;
        var w=w +2*b;
        var h=h +2*b;
        ctx.beginPath();
        ctx.moveTo(cx-w/2-k,cy);
        ctx.lineTo(cx-w/2-rm,cy);
        ctx.arc(cx-w/2-rm, cy+rm, rm, 1.5 * Math.PI, 2.0 * Math.PI, false);
        ctx.arc(cx-w/2+r, cy+rm, r, 3 * Math.PI, 2.5 * Math.PI, true);
        ctx.lineTo(cx+w/2-r, cy+h/2, r, 3 * Math.PI, 2.5 * Math.PI, true);
        ctx.arc(cx+w/2-r, cy+rm, r, 2.5 * Math.PI, 2.0 * Math.PI, true);
        ctx.arc(cx+w/2+rm, cy+rm, rm, 3 * Math.PI, 1.5 * Math.PI, false);
        ctx.lineTo(cx+w/2+k,cy);
        ctx.stroke()
    },
     topline:function(ctx,cx,cy,w,h,r,b,th,dr,side){
        var r=r+b
        var dx=(w+b*2)/2-r
        var dy=(h+b*2)/2-r
        ctx.beginPath();
        if (side=='L'){
            ctx.moveTo(cx-dx,cy-h/2-b-th);
            ctx.arc(cx-dx,cy-dy,r,1.5 * Math.PI,(1.5-dr) * Math.PI, true);
        }else{
            ctx.moveTo(cx+dx,cy-h/2-b-th);
            ctx.arc(cx+dx,cy-dy,r,1.5 * Math.PI, (1.5+dr) * Math.PI, false)
        }
        
        ctx.stroke()
     },
     roundrect:function(ctx,cx,cy,w,h,r){
        var dx=w/2-r
        var dy=h/2-r
        ctx.beginPath();
        ctx.moveTo(cx-dx,cy-h/2);
        ctx.lineTo(cx+dx,cy-h/2);
        ctx.arc(cx+dx, cy-dy, r, 1.5 * Math.PI, 2.0 * Math.PI, false);
        ctx.lineTo(cx+dx+r,cy+dy);
        ctx.arc(cx+dx, cy+dy, r, 2.0 * Math.PI, 2.5 * Math.PI, false);
        ctx.lineTo(cx-dx,cy+dy+r);
        ctx.arc(cx-dx, cy+dy, r, 2.5 * Math.PI, 3.0 * Math.PI, false);
        ctx.lineTo(cx-dx-r,cy-dy);
        ctx.arc(cx-dx, cy-dy, r, 3.0 * Math.PI, 3.5 * Math.PI, false);
        ctx.stroke()
     },
   pigbackground:function(ctx,cx,cy,pig){
        var grad = ctx.createRadialGradient(cx,cy,0,cx,cy,pig.nose.width*2);
        grad.addColorStop(0, 'rgba(255,255,255,1)');
       // grad.addColorStop(0.9, '#019F62');
        grad.addColorStop(1, 'rgba(200,230,255,0)');
        //grad.addColorStop(1, 'rgba(255,255,255,1)');
        ctx.fillStyle = grad;
        ctx.fillRect(0,0,cx*2,cy*2);
   },
   pigtext:function(ctx,cx,cy,pig,s){
       ctx.textAlign= 'center'
       ctx.font=  parseInt(24*s)+"px arial black ";
       ctx.textBaseline = "top";
       ctx.fillStyle = "black";
       ctx.fillText('Genropy', cx, cy+pig.nose.height/2+pig.muzzle.border)
   },
   genropig:function(domnode,kw){
        var context =domnode.getContext('2d');
        var cx= domnode.width/2
        var cy= domnode.height/2
        var s=kw.size/100
        var pig={ nose:{width:90*s,height:70*s,radius:35*s},
                  nostril:{width:25*s,height:25*s,radius:12*s,distance:36*s},
                  muzzle:{border:10*s,sideradius:3*s,sidewidth:10*s},
                  top:{height:10*s,rad:0.3}
        }
        this.pigbackground(context,cx,cy,pig)
        context.lineWidth=6*s;
        context.strokeStyle=kw.strokeStyle;
        context.fillStyle=kw.fillStyle
        this.roundrect(context,cx,cy,pig.nose.width,pig.nose.height,pig.nose.radius)
        this.roundrect(context,cx-pig.nostril.distance/2,cy,pig.nostril.width,pig.nostril.height,pig.nostril.radius)
        this.roundrect(context,cx+pig.nostril.distance/2,cy,pig.nostril.width,pig.nostril.height,pig.nostril.radius)
        this.muzzleline(context,cx,cy,pig.nose.width,pig.nose.height,pig.nose.radius,pig.muzzle.border,pig.muzzle.sidewidth,pig.muzzle.sideradius)
        this.topline(context,cx,cy,pig.nose.width,pig.nose.height,pig.nose.radius,pig.muzzle.border,pig.top.height,pig.top.rad,'L')
        this.topline(context,cx,cy,pig.nose.width,pig.nose.height,pig.nose.radius,pig.muzzle.border,pig.top.height,pig.top.rad,'R')
        //this.pigtext(context,cx,cy,pig,s)
    }
}